File: src/leetcode/javascript/medium/438.找到字符串中所有字母异位词.js
/*
 * @lc app=leetcode.cn id=438 lang=javascript
 *
 * [438] 找到字符串中所有字母异位词
 */

// @lc code=start
/**
 * @param {string} s
 * @param {string} p
 * @return {number[]}
 */
var findAnagrams = function (s, p) {
  const res = []
  const sLen = s.length
  const pLen = p.length
  const needMap = new Map()

  for (const letter of p) {
    if (needMap.has(letter)) {
      needMap.set(letter, needMap.get(letter) + 1)
    } else {
      needMap.set(letter, 1)
    }
  }

  const needLen = needMap.size
  const windowMap = new Map()

  let left = 0
  let right = 0
  let meetedLen = 0

  while (right < sLen) {
    const rightLetter = s[right]
    right++

    if (needMap.has(rightLetter)) {
      if (windowMap.has(rightLetter)) {
        windowMap.set(rightLetter, windowMap.get(rightLetter) + 1)
      } else {
        windowMap.set(rightLetter, 1)
      }

      if (windowMap.get(rightLetter) === needMap.get(rightLetter)) {
        meetedLen++
      }
    }

    while (right - left >= pLen) {
      if (meetedLen === needLen) {
        res.push(left)
      }

      const leftLetter = s[left]
      left++

      if (needMap.has(leftLetter)) {
        if (windowMap.get(leftLetter) === needMap.get(leftLetter)) {
          meetedLen--
        }

        windowMap.set(leftLetter, windowMap.get(leftLetter) - 1)
      }
    }
  }

  return res
}
// @lc code=end

File: src/leetcode/rust/src/medium/438.找到字符串中所有字母异位词.rs
pub fn find_anagrams(s: String, p: String) -> Vec<i32> {
    let mut need_map: HashMap<u8, i32> = HashMap::new();

    for letter in p.as_bytes() {
        need_map.entry(*letter).and_modify(|e| *e += 1).or_insert(1);
    }

    let mut res = Vec::new();
    let need_map_len = need_map.len();
    let mut meeted_count = 0;
    let mut window_map: HashMap<u8, i32> = HashMap::new();
    let (mut start, mut end) = (0, 0);

    while end < s.len() {
        let end_letter = s.as_bytes()[end];
        end += 1;

        if need_map.contains_key(&end_letter) {
            window_map
                .entry(end_letter)
                .and_modify(|e| *e += 1)
                .or_insert(1);

            if window_map.get(&end_letter) == need_map.get(&end_letter) {
                meeted_count += 1;
            }
        }

        while end - start >= p.len() {
            if need_map_len == meeted_count {
                res.push(start as i32);
            }

            let start_letter = s.as_bytes()[start];
            start += 1;

            if need_map.contains_key(&start_letter) {
                if window_map.get(&start_letter) == need_map.get(&start_letter) {
                    meeted_count -= 1;
                }

                window_map.entry(start_letter).and_modify(|e| *e -= 1);
            }
        }
    }

    res
}
